src/IO/H5/Python/ExtractDatFromH5.py

```diff
@@ -0,0 +1,120 @@
+#!/usr/bin/env python
+
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import h5py
+from spectre.IO import H5 as spectre_h5
+import spectre.DataStructures
+import numpy as np
+import os
+import shutil
+
+
+def available_dat_files(h5file, path='/'):
+    subfiles = []
+    for member in h5file.keys():
+        if type(h5file[member]) is h5py._hl.group.Group:
+            subfiles = subfiles + available_dat_files(h5file[member],
+                                                      path + member + '/')
+        elif member.endswith('.dat'):
+            subfiles.append(path + member)
+    return subfiles
+
+
+def extract_dat_files(filename, **kwargs):
+    # Need both of these because it's way easier to list all dat files with h5py
+    # but we still need spectre_h5 to actually get the data
+    h5file_spectre = spectre_h5.H5File(filename, "r")
+    h5file_h5py = h5py.File(filename, "r")
```

Comment:
This looks extremely unsafe... HDF5 has all sorts of issues with multiply opening files, edge cases, etc. Can you make the code so that only one of the libraries has the file open at a time?

Related: what would need to be added to our H5 code that would make it easy to get all this? It seems like we could easily add a `all_subfiles()` function that would list all the available subfiles. It could even take an optional extension and filter on those. I think that's the way to handle this. Almost certainly at some point someone will want to list what's in an HDF5 file from C++.

Reply:
Yes it would be nice to do this with our H5 pybindings. Currently there's a function `groups()` in `IO/H5/File.hpp` that only gets the groups in the root directory. I'm not sure how to extend this to retrieve all subfiles in subgroups, or use a filter to only get Dat files.

Reply:
```cpp
std::vector<std::string> all_files(hid_t file_id, const std:string& group_name) {
  std::vector<std::string> groups = get_group_names(file_id, group_name);
  const range_end = std::remove(groups.begin(), groups.end(), [] (const std::string& t) { return /*check for extension*/; });
  std::vector<std::string> files(range_end, groups.end());
  for (auto it = groups.begin(); it != range_end; ++it) {
    const auto extra_files = all_files(file_id, group);
    files.insert(files.end(), extra_files.begin(), extra_files.end());
  }
  return files;
}
```
Won't work exactly, but should be enough to do it. You don't need anything to do with the HDF5 library directly, you can just use the wrappers :) There might need to be some sort of `detail::OpenGroup` but as long as you have the absolute path then `get_group_names` will take care of it for you. Happy to hop on a call and chat about this if that would be helpful :)